### PR TITLE
*: speed up the operation of "admin check table"

### PR DESCRIPTION
**Pull request by zimulala**
### What problem does this PR solve? 
Reduce the execution time of this operation of "admin check table".

### What is changed and how it works?
compare the result of "select count(*) from table" and "select count(*) from table use index(idx)"

- table count == index count
We will check the records with the `IndexLookUpReader` method of all indexes.

- table count > index count 
We will use the current index's `checkTableRecord` to check the records.

- table count < index count 
We will use the current index's  IndexLookUpReader to check the records.

Using 3 TiKV(8  Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz), 1 PD, 1 TiDB , the servers of TiKV and TiDB are not on a server.

```
CREATE TABLE `x` (
  `a` int(11) NOT NULL,
  `b` varchar(255) DEFAULT NULL,
  `c` blob DEFAULT NULL,
  `d` time DEFAULT NULL,
  `e` double DEFAULT NULL,
  `f` text DEFAULT NULL,
  `g` bit(10) DEFAULT NULL,
  PRIMARY KEY (`a`),
  KEY `d` (`d`),
  KEY `a_2` (`a`,`b`),
  KEY `c` (`c`(10),`d`),
  KEY `e` (`e`,`f`(100)),
  KEY `a_3` (`a`,`b`,`c`(2048)),
  KEY `d_2` (`d`,`e`,`f`(1024),`g`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
```
before vs after
```
mysql> select count(*) from x use index(d);
+----------+
| count(*) |
+----------+
| 10000000 |
+----------+
1 row in set (2.30 sec)

before this PR:
mysql> admin check table x;
More than 3 hours

after this PR:
mysql> admin check table x;
Query OK, 0 rows affected (45.52 sec)
```

Executing this operation with old code takes too long, so the following only tests the new code
```
mysql> select count(*) from x use index(d);
+-----------+
| count(*)  |
+-----------+
| 100000000 |
+—————+
Query OK, 0 rows affected (4 min 27.76 sec)
```

```
CREATE TABLE `lineitem` (
  `L_ORDERKEY` int(11) NOT NULL,
  `L_PARTKEY` int(11) NOT NULL,
  `L_SUPPKEY` int(11) NOT NULL,
  `L_LINENUMBER` int(11) NOT NULL,
  `L_QUANTITY` decimal(15,2) NOT NULL,
  `L_EXTENDEDPRICE` decimal(15,2) NOT NULL,
  `L_DISCOUNT` decimal(15,2) NOT NULL,
  `L_TAX` decimal(15,2) NOT NULL,
  `L_RETURNFLAG` char(1) NOT NULL,
  `L_LINESTATUS` char(1) NOT NULL,
  `L_SHIPDATE` date NOT NULL,
  `L_COMMITDATE` date NOT NULL,
  `L_RECEIPTDATE` date NOT NULL,
  `L_SHIPINSTRUCT` char(25) NOT NULL,
  `L_SHIPMODE` char(10) NOT NULL,
  `L_COMMENT` varchar(44) NOT NULL,
  PRIMARY KEY (`L_ORDERKEY`,`L_LINENUMBER`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin 

mysql> select count(*) from TPCH10.lineitem;
+----------+
| count(*) |
+----------+
| 59986052 |
+----------+
1 row in set (4.69 sec)
tidb> admin check table TPCH10.lineitem;
Query OK, 0 rows affected (1 min 16.92 sec)

mysql> select count(*) from TPCH50.lineitem;
+-----------+
| count(*)  |
+-----------+
| 300005811 |
tidb> admin check table TPCH50.lineitem;
Query OK, 0 rows affected (5 min 50.57 sec)

```

### Check List 

Tests 
 - Unit test

Side effects

 - Possible performance regression

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8572)

### CODE REVIEW

**Comment by winkyao** [comment]
Really impressive.

**Comment by winkyao** [comment]
Please fix ci.

**Comment by zimulala** [comment]
PTAL @winkyao @jackysp

**Comment by jackysp** [comment]
OK, great work! I need more time to review it.

**Comment by zimulala** [comment]
PTAL @winkyao @jackysp @crazycs520 @XuHuaiyu

File: executor/executor.go [hunk 1]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
+	cols := src.schema.Columns
+	retFieldTypes := make([]*types.FieldType, len(cols))
+	for i := range cols {
+		retFieldTypes[i] = cols[i].RetType
+	}
+	chk := chunk.New(retFieldTypes, e.initCap, e.maxChunkSize)
+	iter := chunk.NewIterator4Chunk(chk)
+
+	var err error
+	handles := make(map[int64]struct{}, 1024)
```
Comment (jackysp) [inline executor/executor.go]: Why use 1024 here?

File: executor/executor.go [hunk 2]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
-		if err != nil {
-			log.Warnf("%v error:%v", t.Name, errors.ErrorStack(err))
-			if admin.ErrDataInConsistent.Equal(err) {
-				return ErrAdminCheckTable.GenWithStack("%v err:%v", t.Name, err)
+
+		for chunkRow := iter.Begin(); chunkRow != iter.End(); chunkRow = iter.Next() {
+			datum := chunkRow.GetDatum(0, retFieldTypes[0])
+			h := datum.GetInt64()
+			if _, ok := handles[h]; ok {
+				err = errors.Errorf("admin check table %s, index %s handle %d more than one", e.tblInfo.Name, e.indices[num].Meta().Name, h)
```
Comment (jackysp) [inline executor/executor.go]:
```suggestion
				err = errors.Errorf("admin check table %s, index %s, at least two indices have the same handle %d", e.tblInfo.Name, e.indices[num].Meta().Name, h)
```

File: executor/executor.go [hunk 3]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
+	chk := chunk.New(retFieldTypes, e.initCap, e.maxChunkSize)
+	iter := chunk.NewIterator4Chunk(chk)
+
+	var err error
+	handles := make(map[int64]struct{}, 1024)
+	for {
+		err = src.Next(ctx, chk)
 		if err != nil {
-			return errors.Trace(err)
+			break
```
Comment (jackysp) [inline executor/executor.go]: Why not return the error?

File: executor/executor.go [hunk 4]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
+	for _, idx := range e.indices {
+		idxNames = append(idxNames, idx.Meta().Name.O)
+	}
+	greater, idxOffset, err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tblInfo.Name.O, idxNames)
+	if err != nil {
+		log.Warnf("check table %v, greater %v index %s err: %v", e.tblInfo.Name, greater, idxNames[idxOffset], errors.ErrorStack(err))
+		tbl := e.srcs[idxOffset].table
+		txn := e.ctx.Txn(true)
+		switch greater {
+		case admin.IdxCntGreater:
```
Comment (jackysp) [inline executor/executor.go]: I think an if/else is good enough for this switch/case.

File: util/admin/admin.go [hunk 5]
```diff
@@ -248,28 +248,46 @@ func getCount(ctx sessionctx.Context, sql string) (int64, error) {
...
+	InvalidGreater byte = 0
+	// TblCntGreater means that the number of table rows is more than the number of index rows.
+	TblCntGreater byte = 1
+	// IdxCntGreater means that the number of index rows is more than the number of table rows.
+	IdxCntGreater byte = 2
+)
+
 // CheckIndicesCount compares indices count with table count.
 // It returns nil if the count from the index is equal to the count from the table columns,
 // otherwise it returns an error with a different information.
```
Comment (jackysp) [inline util/admin/admin.go]: It's better to describe the return values in the comment.

File: util/admin/admin.go [hunk 6]
```diff
@@ -248,28 +248,46 @@ func getCount(ctx sessionctx.Context, sql string) (int64, error) {
...
+		return 0, 0, errors.Trace(err)
 	}
-	for _, idx := range indices {
+
+	for i, idx := range indices {
 		sql = fmt.Sprintf("SELECT COUNT(*) FROM `%s`.`%s` USE INDEX(`%s`)", dbName, tableName, idx)
 		idxCnt, err := getCount(ctx, sql)
 		if err != nil {
-			return errors.Trace(err)
+			return InvalidGreater, i, errors.Trace(err)
```
Comment (jackysp) [inline util/admin/admin.go]: If we check the error first, I think we could avoid `InvalidGreater`.

File: executor/executor.go [hunk 7]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
 }
 
-func (e *CheckTableExec) doCheckPartitionedTable(tbl table.PartitionedTable) error {
-	info := tbl.Meta().GetPartitionInfo()
-	for _, def := range info.Definitions {
-		pid := def.ID
-		partition := tbl.GetPartition(pid)
-		if err := e.doCheckTable(partition); err != nil {
+// Next implements the Executor Next interface.
+func (e *CheckTableExec) Next(ctx context.Context, chk *chunk.Chunk) error {
```
Comment (jackysp) [inline executor/executor.go]: Seems we only check the index handle value except for the number of rows on the table is greater than the index.
I think it is better to check the index values.

File: executor/executor.go [hunk 8]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
+	greater, idxOffset, err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tblInfo.Name.O, idxNames)
+	if err != nil {
+		log.Warnf("check table %v, greater %v index %s err: %v", e.tblInfo.Name, greater, idxNames[idxOffset], errors.ErrorStack(err))
+		tbl := e.srcs[idxOffset].table
+		txn := e.ctx.Txn(true)
+		switch greater {
+		case admin.IdxCntGreater:
+			err = e.checkIndexHandle(ctx, idxOffset, e.srcs[idxOffset])
+		case admin.TblCntGreater:
+			err = e.checkTableRecord(txn, tbl, e.indices[idxOffset])
```
Comment (jackysp) [inline executor/executor.go]: If the count of rows on the table is greater, the `admin check table` will run as before?

Comment (zimulala) [reply executor/executor.go]: No. Now we only call `CheckRecordAndIndex` here. Before this PR, we call `checkIndexAndRecord` and `CheckRecordAndIndex`.

File: executor/executor.go (hunk 7 above)
Comment (zimulala) [reply executor/executor.go]: I have asked your opinions before, you feel that you can accept not check index value. If we really need it, we can handle it in the next PR.

File: executor/executor.go (hunk 1 above)
Comment (zimulala) [reply executor/executor.go]: Do you have any suggestions?

**Comment by zimulala** [comment]
PTAL @jackysp

File: executor/executor.go (hunk 7 above)
Comment (jackysp) [reply executor/executor.go]: I used to think your question is about the error message with the handle only or with the values. I think it is my fault to misunderstand your question.

File: executor/executor.go (hunk 1 above)
Comment (jackysp) [reply executor/executor.go]: I don't have any suggestions until I know why it is 1024 here.

File: executor/executor.go (hunk 3 above)
Comment (jackysp) [reply executor/executor.go]: ok

File: executor/distsql.go [hunk 9]
```diff
@@ -334,7 +336,7 @@ type IndexLookUpExecutor struct {
 	memTracker *memory.Tracker
 
 	// isCheckOp is used to determine whether we need to check the consistency of the index data.
-	isCheckOp bool
+	checkIndexValue
```
Comment (crazycs520) [inline executor/distsql.go]: How about use `*checkIndexValue` and remove `checkIndexValue` filed `isCheckOp`, then use `w.checkIndexValue != nil` to check.

BYW, need to update the comment in line#339, 622, 738.

File: executor/distsql.go [hunk 10]
```diff
@@ -652,29 +671,40 @@ func (w *indexWorker) fetchHandles(ctx context.Context, result distsql.SelectRes
...
+func (w *indexWorker) extractTaskHandles(ctx context.Context, chk *chunk.Chunk, idxResult distsql.SelectResult) (
+	handles []int64, retChk *chunk.Chunk, err error) {
+	if w.isCheckOp {
+		chk = chunk.NewChunkWithCapacity(w.tps, w.maxChunkSize)
+	}
 	handles = make([]int64, 0, w.batchSize)
 	for len(handles) < w.batchSize {
 		chk.SetRequiredRows(w.batchSize-len(handles), w.maxChunkSize)
-		err = idxResult.Next(ctx, chk)
+		err = errors.Trace(idxResult.Next(ctx, chk))
```
Comment (crazycs520) [inline executor/distsql.go]: Why add `errors.Trace` here but not in line#684?

File: executor/distsql.go [hunk 11]
```diff
@@ -652,29 +671,40 @@ func (w *indexWorker) fetchHandles(ctx context.Context, result distsql.SelectRes
...
-			return handles, err
+			return handles, nil, err
 		}
 		if chk.NumRows() == 0 {
-			return handles, nil
+			return handles, retChk, nil
 		}
 		for i := 0; i < chk.NumRows(); i++ {
-			handles = append(handles, chk.GetRow(i).GetInt64(0))
+			h := chk.GetRow(i).GetInt64(chk.NumCols() - 1)
```
Comment (crazycs520) [inline executor/distsql.go]: How about calculate `chk.NumCols() - 1` outside the for loop.

File: planner/core/planbuilder.go [hunk 12]
```diff
@@ -15,7 +15,9 @@ package core
 
 import (
 	"bytes"
+	"context"
 	"fmt"
+	"go.uber.org/zap"
```
Comment (crazycs520) [inline planner/core/planbuilder.go]: Put this in the right place.

File: executor/admin_test.go [hunk 13]
```diff
@@ -399,7 +399,98 @@ func (s *testSuite2) TestAdminCleanupIndexMore(c *C) {
...
+	tblName := model.NewCIStr("admin_test")
+	tbl, err := is.TableByName(dbName, tblName)
+	c.Assert(err, IsNil)
+	tblInfo := tbl.Meta()
+	idxInfo := tblInfo.Indices[1]
+	indexOpr := tables.NewIndex(tblInfo.ID, tblInfo, idxInfo)
+	sc := s.ctx.GetSessionVars().StmtCtx
+	tk.Se.GetSessionVars().IndexLookupSize = 3
+	tk.Se.GetSessionVars().MaxChunkSize = 3
+
```
Comment (crazycs520) [inline executor/admin_test.go]: I found a bug, Please add below test.

Below test is for handle < 0;
```go
	tk.MustExec("insert into admin_test values (-1,-1,'')")
	txn, err := s.store.Begin()
	c.Assert(err, IsNil)
	err = indexOpr.Delete(sc, txn, types.MakeDatums(-1), -1, nil)
	c.Assert(err, IsNil)
	err = txn.Commit(context.Background())
	c.Assert(err, IsNil)
	_, err = tk.Exec("admin check table admin_test")
	c.Assert(err, NotNil)
	tk.MustExec("delete from admin_test where c1=-1")
```

There is a bug in [util/admin/admin.go line#464](https://github.com/pingcap/tidb/pull/8572/files#diff-25fb496eed984a134a4a4c98e8b687acL446), it should be `startKey := t.RecordKey(math.MinInt64)`

Comment (zimulala) [reply executor/admin_test.go]: Done.

File: executor/distsql.go (hunk 9 above)
Comment (zimulala) [reply executor/distsql.go]: `checkIndexValue.genExprs` is assigned in a shared function, so we can't use `w.checkIndexValue != nil` to check if we need to check the index value. So I don't update it.

**Comment by zimulala** [comment]
PTAL @crazycs520

File: go.mod [hunk 14]
```diff
@@ -31,6 +31,7 @@ require (
 	github.com/modern-go/reflect2 v1.0.1 // indirect
 	github.com/montanaflynn/stats v0.0.0-20180911141734-db72e6cae808 // indirect
 	github.com/myesui/uuid v1.0.0 // indirect
+	github.com/ngaut/log v0.0.0-20180314031856-b8e36e7ba5ac
```
Comment (jackysp) [inline go.mod]: The old log comes back?

Comment (zimulala) [reply go.mod]: No, I will remove it.

File: executor/builder.go [hunk 15]
```diff
@@ -320,12 +320,61 @@ func (b *executorBuilder) buildCheckIndex(v *plannercore.CheckIndex) Executor {
 	return e
 }
 
+func buildIndexLookUpChecker(b *executorBuilder, readerPlan *plannercore.PhysicalIndexLookUpReader,
+	readerExec *IndexLookUpExecutor) {
+	readerExec.keepOrder = true
```
Comment (crazycs520) [inline executor/builder.go]: Why need to keep order?

File: executor/distsql.go [hunk 16]
```diff
@@ -652,29 +671,41 @@ func (w *indexWorker) fetchHandles(ctx context.Context, result distsql.SelectRes
 	}
 }
 
-func (w *indexWorker) extractTaskHandles(ctx context.Context, chk *chunk.Chunk, idxResult distsql.SelectResult) (handles []int64, err error) {
+func (w *indexWorker) extractTaskHandles(ctx context.Context, chk *chunk.Chunk, idxResult distsql.SelectResult) (
+	handles []int64, retChk *chunk.Chunk, err error) {
+	if w.isCheckOp {
```
Comment (crazycs520) [inline executor/distsql.go]: Can we move line#676~line#678 to line#647 to reduce reallocate memory?
```go
if w.isCheckOp {
		chk = chunk.NewChunkWithCapacity(w.tps, w.maxChunkSize)
	} else {
		chk = chunk.NewChunkWithCapacity([]*types.FieldType{types.NewFieldType(mysql.TypeLonglong)}, w.idxLookup.maxChunkSize)
	}
``

File: executor/distsql.go [hunk 17]
```diff
@@ -652,29 +671,41 @@ func (w *indexWorker) fetchHandles(ctx context.Context, result distsql.SelectRes
...
 		if chk.NumRows() == 0 {
-			return handles, nil
+			return handles, retChk, nil
 		}
 		for i := 0; i < chk.NumRows(); i++ {
-			handles = append(handles, chk.GetRow(i).GetInt64(0))
+			h := chk.GetRow(i).GetInt64(handleOffset)
+			handles = append(handles, h)
+		}
+		if w.isCheckOp {
```
Comment (crazycs520) [inline executor/distsql.go]: This logic has nothing to do with this function, how about move this logic to line#662.
and change
```go
retChk = chunk.NewChunkWithCapacity(w.tps, w.maxChunkSize)
```
to 
```go
// no need to allocate max chunk size.
retChk = chunk.NewChunkWithCapacity(w.tps, chk.NumRows())
```

File: executor/distsql.go [hunk 18]
```diff
@@ -652,29 +671,41 @@ func (w *indexWorker) fetchHandles(ctx context.Context, result distsql.SelectRes
...
+			if retChk == nil {
+				retChk = chunk.NewChunkWithCapacity(w.tps, w.maxChunkSize)
```
Comment (crazycs520) [inline executor/distsql.go]:
```go
// no need to allocate max chunk size.
retChk = chunk.NewChunkWithCapacity(w.tps, w.batchSize)
```

File: executor/distsql.go [hunk 19]
```diff
@@ -737,6 +770,80 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
...
+				vals1[i].SetBytes(val1.GetBytes())
+				vals2[i].SetBytes(val2.GetBytes())
+			}
+		}
+	}
+}
+
+func (w *tableWorker) compareData(ctx context.Context, task *lookupTableTask, tableReader Executor) error {
+	for {
+		chk := newFirstChunk(tableReader)
```
Comment (crazycs520) [inline executor/distsql.go]: move this outside this for loop?

File: executor/distsql.go [hunk 20]
```diff
@@ -737,6 +770,80 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
...
+
+		tblReaderExec := tableReader.(*TableReaderExecutor)
+		i := 0
+		iter := chunk.NewIterator4Chunk(chk)
+		for row := iter.Begin(); row != iter.End(); row = iter.Next() {
+			handle := row.GetInt64(row.Len() - 1)
+			offset := task.indexOrder[handle]
+			delete(task.indexOrder, handle)
+			idxRow := task.idxRows.GetRow(offset)
+			vals := make([]types.Datum, 0, len(w.cols))
```
Comment (crazycs520) [inline executor/distsql.go]: ditto, move this outside the for loop (before line#791) to avoid allocating memory?

File: executor/distsql.go [hunk 21]
```diff
@@ -737,6 +770,80 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
...
+					vals = append(vals, row.GetDatum(j, &col.FieldType))
+				}
+			}
+			vals = tables.TruncateIndexValuesIfNeeded(w.tbl.Meta(), w.idxInfo, vals)
+			for j, val := range vals {
+				col := w.cols[j]
+				tp := &col.FieldType
+				ret := chunk.Compare(idxRow, j, &val)
+				if ret != 0 {
+					return errors.Errorf("handle %#v, index:%#v != record:%#v", handle, idxRow.GetDatum(j, tp), val)
```
Comment (crazycs520) [inline executor/distsql.go]: Can we print out the col.Name?

File: executor/distsql.go [hunk 22]
```diff
@@ -737,6 +770,80 @@ func (w *tableWorker) pickAndExecTask(ctx context.Context) {
...
+		if chk.NumRows() == 0 {
+			for h := range task.indexOrder {
+				idxRow := task.idxRows.GetRow(task.indexOrder[h])
+				return errors.Errorf("handle %#v, index:%#v != record:%#v", h, idxRow.GetDatum(0, w.tps[0]), nil)
+			}
+			break
+		}
+
+		tblReaderExec := tableReader.(*TableReaderExecutor)
+		i := 0
```
Comment (crazycs520) [inline executor/distsql.go]: `i` look like useless?

File: executor/admin_test.go [hunk 23]
```diff
@@ -399,7 +399,98 @@ func (s *testSuite2) TestAdminCleanupIndexMore(c *C) {
...
+	r := tk.MustQuery("admin recover index admin_test c2")
+	r.Check(testkit.Rows("1 6"))
+	tk.MustExec("admin check table admin_test")
+
+	// Add one row of index.
+	// Table count < index count.
+	// Index c2 has two more values ​​than table data: 21, 13.
+	txn, err = s.store.Begin()
+	c.Assert(err, IsNil)
+	_, err = indexOpr.Create(s.ctx, txn, types.MakeDatums(21), 1)
```
Comment (crazycs520) [inline executor/admin_test.go]: The handle is 1, and has correlative record, 
Could you add test for Table count < index count, and the handle stored in the index doesn't have the correlative record. Then the admin check table error should like this `"handle 12, index:types.Datum{k:0x1, collation:0x0, decimal:0x0, length:0x0, i:13, b:[]uint8(nil), x:interface {}(nil)} != record:<nil>")`, 
The key point here is  `record:<nil>`

File: executor/executor.go [hunk 24]
```diff
@@ -455,75 +455,132 @@ func getTableName(is infoschema.InfoSchema, id int64) string {
...
 		}
-		if err != nil {
-			logutil.Logger(ctx).Warn("check table failed", zap.String("tableName", t.Name.O), zap.Error(err))
-			if admin.ErrDataInConsistent.Equal(err) {
-				return ErrAdminCheckTable.GenWithStack("%v err:%v", t.Name, err)
+		for chunkRow := iter.Begin(); chunkRow != iter.End(); chunkRow = iter.Next() {
+			datum := chunkRow.GetDatum(0, retFieldTypes[0])
+			h := datum.GetInt64()
+			if _, ok := handles[h]; ok {
+				err = errors.Errorf("admin check table %s, index %s, at least two indices have the same handle %d more than one",
```
Comment (crazycs520) [inline executor/executor.go]: Could you add a test case to let `admin check` return this error?

File: executor/executor.go [hunk 25]
```diff
@@ -455,75 +455,132 @@ func getTableName(is infoschema.InfoSchema, id int64) string {
...
+	greater, idxOffset, err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tblInfo.Name.O, idxNames)
+	if err != nil {
+		tbl := e.srcs[idxOffset].table
+		if greater == admin.IdxCntGreater {
+			err = e.checkIndexHandle(ctx, idxOffset, e.srcs[idxOffset])
+		} else if greater == admin.TblCntGreater {
+			err = e.checkTableRecord(tbl, idxOffset)
+		}
+		if err != nil && admin.ErrDataInConsistent.Equal(err) {
+			return ErrAdminCheckTable.GenWithStack("%v err:%v", tbl.Meta().Name, err)
```
Comment (crazycs520) [inline executor/executor.go]: Why you add this error check here?  We can contain the table name directly when do `admin.CheckRecordAndIndex`,  
`return ErrAdminCheckTable.GenWithStack("%v err:%v", tbl.Meta().Name, err)` 
here will lost the original stack info and then I don't know where the error returns from.

Comment (zimulala) [reply executor/executor.go]: I think it wants to unify the error message. It the original code. And the function calls `GenWithStack` it should have the original stack info.

File: executor/admin_test.go (hunk 23 above)
Comment (zimulala) [reply executor/admin_test.go]: Done.

File: executor/builder.go (hunk 15 above)
Comment (zimulala) [reply executor/builder.go]: To save the handle offsets. But now I replace it with `isCheckOp` .

**Comment by zimulala** [comment]
PTAL @crazycs520

File: planner/core/planbuilder.go [hunk 26]
```diff
@@ -569,42 +571,7 @@ func (b *PlanBuilder) buildCheckIndex(dbName model.CIStr, as *ast.AdminStmt) (Pl
...
-	// It's double read case.
-	ts := PhysicalTableScan{Columns: columns, Table: is.Table}.Init(b.ctx)
-	ts.SetSchema(is.dataSourceSchema)
-	cop.tablePlan = ts
-	is.initSchema(id, idx, true)
-	t := finishCopTask(b.ctx, cop)
-
-	rootT := t.(*rootTask)
-	return rootT.p, nil
+	return b.buildPhysicalIndexLookUpReader(ctx, dbName, tbl, idx, 1)
```
Comment (AilinKid) [inline planner/core/planbuilder.go]: As I know the last parameter is position info of index start from 0.
And this parameter will be passed from `buildPhysicalIndexLookUpReader()` to `is.initSchema()`, which eventually has no use of this parameter.

File: planner/core/planbuilder.go [hunk 27]
```diff
@@ -690,43 +657,170 @@ func (b *PlanBuilder) buildAdmin(ctx context.Context, as *ast.AdminStmt) (Plan,
...
+	tbl := as.Tables[0]
+	p := &CheckTable{
+		DBName:  tbl.Schema.O,
+		TblInfo: tbl.TableInfo,
+	}
+
+	mockTablePlan := LogicalTableDual{}.Init(b.ctx)
+	tableInfo := as.Tables[0].TableInfo
+	schema := expression.TableInfo2SchemaWithDBName(b.ctx, tbl.Schema, tableInfo)
+	mockTablePlan.SetSchema(schema)
```
Comment (AilinKid) [inline planner/core/planbuilder.go]: It's seems like we have defined a mockTablePlan here without use. 
Similar thing in `buildPhysicalIndexLookUpReader`—— downstream of the call chain.

File: planner/core/planbuilder.go [hunk 28]
```diff
@@ -690,43 +657,170 @@ func (b *PlanBuilder) buildAdmin(ctx context.Context, as *ast.AdminStmt) (Plan,
...
+	rootT := t.(*rootTask)
+	return rootT.p, nil
+}
+
+func (b *PlanBuilder) buildPhysicalIndexLookUpReaders(ctx context.Context, dbName model.CIStr, tbl table.Table) ([]Plan, []table.Index, error) {
+	tblInfo := tbl.Meta()
+	// get index information
+	indices := make([]table.Index, 0, len(tblInfo.Indices))
+	indexLookUpReaders := make([]Plan, 0, len(tblInfo.Indices))
+	for i, idx := range tbl.Indices() {
```
Comment (winkyao) [inline planner/core/planbuilder.go]: To decrease indent:
```
for i, idx := range tbl.Indices() {
		idxInfo := idx.Meta()
		if idxInfo.State != model.StatePublic {
			logutil.Logger(context.Background()).Info("build physical index lookup reader, the index isn't public",
				zap.String("index", idxInfo.Name.O), zap.Stringer("state", idxInfo.State), zap.String("table", tblInfo.Name.O))
			continue
		} 

		indices = append(indices, idx)
		reader, err := b.buildPhysicalIndexLookUpReader(ctx, dbName, tbl, idxInfo, i)
		if err != nil {
			return nil, nil, err
		}
		indexLookUpReaders = append(indexLookUpReaders, reader)
	}
```

File: planner/core/planbuilder.go [hunk 29]
```diff
@@ -690,43 +657,170 @@ func (b *PlanBuilder) buildAdmin(ctx context.Context, as *ast.AdminStmt) (Plan,
...
+			}
+			genColumnID := model.TableColumnID{TableID: tblInfo.ID, ColumnID: col.ID}
+			if expr, ok := genExprs[genColumnID]; ok {
+				cols := expression.ExtractColumns(expr)
+				genCols = append(genCols, cols...)
 			}
+		}
+	}
+	tblSchema := schema.Clone()
+	for _, col := range genCols {
```
Comment (winkyao) [inline planner/core/planbuilder.go]: I don't understand what is doing here? Could you add some comments here?

File: planner/core/planbuilder.go [hunk 30]
```diff
@@ -690,43 +657,170 @@ func (b *PlanBuilder) buildAdmin(ctx context.Context, as *ast.AdminStmt) (Plan,
...
+		Index:            idx,
+		dataSourceSchema: schema,
+		Ranges:           ranger.FullRange(),
+		KeepOrder:        false,
+		GenExprs:         genExprs,
+	}.Init(b.ctx)
+	is.stats = property.NewSimpleStats(0)
+	cop := &copTask{indexPlan: is}
+	// It's double read case.
+	ts := PhysicalTableScan{Columns: tblColumns, Table: is.Table}.Init(b.ctx)
```
Comment (winkyao) [inline planner/core/planbuilder.go]: And could you use a specific function to build PhysicalTableScan, and another for building PhysicalIndexScan? I think that will make the code more readable.

File: planner/core/planbuilder.go [hunk 31]
```diff
@@ -690,43 +657,170 @@ func (b *PlanBuilder) buildAdmin(ctx context.Context, as *ast.AdminStmt) (Plan,
 	return ret, nil
 }
 
-func (b *PlanBuilder) buildAdminCheckTable(ctx context.Context, as *ast.AdminStmt) (*CheckTable, error) {
-	p := &CheckTable{Tables: as.Tables}
-	p.GenExprs = make(map[model.TableColumnID]expression.Expression, len(p.Tables))
+func (b *PlanBuilder) buildPhysicalIndexLookUpReader(ctx context.Context, dbName model.CIStr, tbl table.Table, idx *model.IndexInfo, id int) (Plan, error) {
+	tblInfo := tbl.Meta()
+	columns := make([]*model.ColumnInfo, 0, len(idx.Columns))
```
Comment (winkyao) [inline planner/core/planbuilder.go]: May be `columns` --> `indexReaderCols` is better?

File: executor/builder.go [hunk 32]
```diff
@@ -322,12 +322,60 @@ func (b *executorBuilder) buildCheckIndex(v *plannercore.CheckIndex) Executor {
 	return e
 }
 
+func buildIndexLookUpChecker(b *executorBuilder, readerPlan *plannercore.PhysicalIndexLookUpReader,
```
Comment (winkyao) [inline executor/builder.go]: Please add some comments for this function.

File: executor/builder.go [hunk 33]
```diff
@@ -322,12 +322,60 @@ func (b *executorBuilder) buildCheckIndex(v *plannercore.CheckIndex) Executor {
...
+	readerExec *IndexLookUpExecutor) {
+	is := readerPlan.IndexPlans[0].(*plannercore.PhysicalIndexScan)
+	readerExec.dagPB.OutputOffsets = make([]uint32, 0, len(is.Index.Columns))
+	for i := 0; i <= len(is.Index.Columns); i++ {
+		readerExec.dagPB.OutputOffsets = append(readerExec.dagPB.OutputOffsets, uint32(i))
+	}
+	// set tps
```
Comment (winkyao) [inline executor/builder.go]: `set tps` has no information.

File: executor/distsql.go [hunk 34]
```diff
@@ -348,6 +353,15 @@ type IndexLookUpExecutor struct {
 	colLens         []int
 }
 
+type checkIndexValue struct {
+	isCheckOp bool
+	tps       []*types.FieldType
+	tbl       table.Table
```
Comment (winkyao) [inline executor/distsql.go]: These fields are duplicated with IndexLookUpExecutor's fields, why not just use the fields in IndexLookUpExecutor?

File: executor/executor.go [hunk 35]
```diff
@@ -455,75 +455,119 @@ func getTableName(is infoschema.InfoSchema, id int64) string {
...
-				return ErrAdminCheckTable.GenWithStack("%v err:%v", t.Name, err)
-			}
+		return errors.Trace(err)
+	}
 
-			return errors.Errorf("%v err:%v", t.Name, err)
-		}
+	// The number of table rows is equal to the number of index rows.
+	for i := range e.srcs {
+		go func(num int) {
```
Comment (winkyao) [inline executor/executor.go]: Please use util.WithRecovery.

File: executor/executor.go [hunk 36]
```diff
@@ -419,56 +424,128 @@ func (e *CheckTableExec) Open(ctx context.Context) error {
...
+	if e.done || len(e.srcs) == 0 {
+		return nil
+	}
+	defer func() { e.done = true }()
+
+	idxNames := make([]string, 0, len(e.indices))
+	for _, idx := range e.indices {
+		idxNames = append(idxNames, idx.Meta().Name.O)
+	}
+	greater, idxOffset, err := admin.CheckIndicesCount(e.ctx, e.dbName, e.tblInfo.Name.O, idxNames)
```
[the remark this answers is not in the document]
Comment (zimulala) [reply executor/executor.go]: If we don't check the count, we don't know which record count is greater. So we need to check records twice. We need to `checkIndexHandle` and `checkTableRecord` for each index.

**Comment by winkyao** [comment]
@zimulala Please address comments

File: planner/core/planbuilder.go (hunk 26 above)
Comment (zimulala) [reply planner/core/planbuilder.go]: Yes, it doesn't be used. So I think this value has no meaning.

**Comment by zimulala** [comment]
PTAL @winkyao @AilinKid

**Comment by AilinKid** [comment]
This pr is too long, need time to deliberate. by now it's LGTM

File: executor/distsql.go [hunk 37]
```diff
@@ -353,6 +358,14 @@ type IndexLookUpExecutor struct {
 	colLens         []int
 }
 
+type checkIndexValue struct {
+	isCheckOp  bool
+	idxColTps  []*types.FieldType
+	idxInfo    *model.IndexInfo
```
Comment (winkyao) [inline executor/distsql.go]: Could we use `IndexLookUpExecutor.index` field?

File: executor/distsql.go (hunk 9 above)
Comment (winkyao) [reply executor/distsql.go]: Maybe use pointer is better? At most cases, this checkIndexValue field is not necessary.

File: executor/distsql.go [hunk 38]
```diff
@@ -348,6 +353,15 @@ type IndexLookUpExecutor struct {
...
 
+type checkIndexValue struct {
+	isCheckOp bool
+	tps       []*types.FieldType
+	tbl       table.Table
+	idxInfo   *model.IndexInfo
+	cols      []*table.Column
+	genExprs  map[model.TableColumnID]expression.Expression
+}
+
```
[the remark this answers is not in the document]
Comment (zimulala) [reply executor/distsql.go]: I updated the field names.

**Pull request merged**